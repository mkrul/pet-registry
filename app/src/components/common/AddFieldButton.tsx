import React from "react";
import { Button } from "@mui/material";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faPlus } from "@fortawesome/free-solid-svg-icons";
import { commonInputStyles } from "../../styles/commonStyles";

interface AddFieldButtonProps {
  onClick: () => void;
  disabled?: boolean;
  label?: string;
  testId?: string;
}

export const AddFieldButton: React.FC<AddFieldButtonProps> = ({
  onClick,
  disabled = false,
  label = "ADD FIELD",
  testId = "add-field-button"
}) => {
  return (
    <Button
      onClick={onClick}
      disabled={disabled}
      color="primary"
      variant="text"
      sx={commonInputStyles}
      data-testid={testId}
    >
      <div className="flex items-center">
        <FontAwesomeIcon icon={faPlus} className="mr-2 mb-[3px]" />
        <span>{label}</span>
      </div>
    </Button>
  );
};
