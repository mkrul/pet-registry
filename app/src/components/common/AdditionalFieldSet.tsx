import React from "react";
import { Collapse } from "@mui/material";

interface AdditionalFieldSetProps {
  label: string;
  children: React.ReactNode;
  show?: boolean;
}

export const AdditionalFieldSet: React.FC<AdditionalFieldSetProps> = ({
  label,
  children,
  show = true
}) => {
  return (
    <Collapse in={show} timeout={300}>
      <div>
        <label className="text-lg font-medium text-gray-900 block mb-2">{label}</label>
        <div className="flex-grow">{children}</div>
      </div>
    </Collapse>
  );
};
