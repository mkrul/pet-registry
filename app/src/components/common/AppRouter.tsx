import { Navigate, Route, Routes } from "react-router-dom";
import NavBar from "../shared/Navbar";
import ReportIndexPage from "../../pages/reports/ReportsIndexPage";
import ReportNewPage from "../../pages/reports/ReportNewPage";
import ReportShowPage from "../../pages/reports/ReportShowPage";
import Footer from "../shared/Footer";
import PrivateRoute from "./PrivateRoute";
import LoginPage from "../../pages/auth/LoginPage";
import OAuthHandler from "../auth/OAuthHandler";
import ScrollToTop from "./ScrollToTop";

const AppRouter = () => {
  return (
    <div>
      <NavBar />
      <ScrollToTop />
      <Routes>
        {/* OAuth Callback Route */}
        <Route path="/auth/callback" element={<OAuthHandler />} />

        {/* Public Routes */}
        <Route path="/login" element={<LoginPage />} />
        <Route path="/" element={<ReportIndexPage />} />
        <Route path="/reports/:id" element={<ReportShowPage />} />

        {/* Protected Routes */}
        <Route element={<PrivateRoute />}>
          <Route path="/reports/new" element={<ReportNewPage />} />
        </Route>

        {/* Fallback Route */}
        <Route path="*" element={<Navigate to="/" />} />
      </Routes>
      <Footer />
    </div>
  );
};

export default AppRouter;
