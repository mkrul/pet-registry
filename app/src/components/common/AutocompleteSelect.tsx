import React from "react";
import { Autocomplete, TextField, FormControl } from "@mui/material";
import { SxProps } from "@mui/material";

interface AutocompleteSelectProps {
  name: string;
  value: string;
  onChange: (e: React.ChangeEvent<HTMLSelectElement>) => void;
  options: string[];
  placeholder: string;
  disabled?: boolean;
  sx?: SxProps;
  size?: "small" | "medium";
}

export const AutocompleteSelect: React.FC<AutocompleteSelectProps> = ({
  name,
  value,
  onChange,
  options,
  placeholder,
  disabled = false,
  sx,
  size = "small"
}) => {
  const handleChange = (_: any, newValue: string | null) => {
    onChange({
      target: { name, value: newValue || "" }
    } as React.ChangeEvent<HTMLSelectElement>);
  };

  return (
    <FormControl fullWidth size={size}>
      <Autocomplete
        value={value}
        onChange={handleChange}
        options={options}
        disabled={disabled}
        size={size}
        clearable
        slotProps={{
          listbox: {
            style: { maxHeight: 200 }
          }
        }}
        renderInput={params => (
          <TextField
            {...params}
            placeholder={placeholder}
            variant="outlined"
            size={size}
            sx={sx}
          />
        )}
      />
    </FormControl>
  );
};

export default AutocompleteSelect;
