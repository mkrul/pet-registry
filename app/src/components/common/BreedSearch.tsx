import React, { useMemo } from "react";
import { TextField, Autocomplete, FormControl } from "@mui/material";
import { getBreedsBySpecies } from "../../lib/reports/breedList";

interface BreedSearchProps {
  species: "dog" | "cat";
  value: string;
  onChange: (breed: string) => void;
  excludeBreeds?: string[];
  required?: boolean;
  disabled?: boolean;
}

export const BreedSearch: React.FC<BreedSearchProps> = ({
  species,
  value,
  onChange,
  excludeBreeds = [],
  required = false,
  disabled = false
}) => {
  const breedOptions = useMemo(() => {
    const breeds = species ? getBreedsBySpecies(species) : [];
    return breeds.filter(breed => !excludeBreeds.includes(breed));
  }, [species, excludeBreeds]);

  return (
    <FormControl fullWidth>
      <Autocomplete
        value={value}
        onChange={(_, newValue) => onChange(newValue || "")}
        options={breedOptions}
        disabled={disabled || !species}
        renderInput={params => (
          <TextField
            {...params}
            required={required}
            label="Breed"
            variant="outlined"
            sx={{
              "& .MuiOutlinedInput-root": {
                backgroundColor: "white !important"
              }
            }}
          />
        )}
      />
    </FormControl>
  );
};

export default BreedSearch;
