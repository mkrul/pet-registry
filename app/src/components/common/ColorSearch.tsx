import React from "react";
import { TextField, Autocomplete } from "@mui/material";
import { getColorOptions } from "../../lib/reports/colorList";

interface ColorSearchProps {
  value: string;
  onChange: (value: string) => void;
  excludeColors?: (string | null)[];
  disabled?: boolean;
  size?: "small" | "medium";
}

export const ColorSearch: React.FC<ColorSearchProps> = ({
  value,
  onChange,
  excludeColors = [],
  disabled = false,
  size = "small"
}) => {
  const colorOptions = React.useMemo(() => {
    const colors = getColorOptions();
    return colors.filter(color => !excludeColors.includes(color));
  }, [excludeColors]);

  return (
    <Autocomplete
      value={value}
      onChange={(_, newValue) => onChange(newValue || "")}
      options={colorOptions}
      disabled={disabled}
      size={size}
      renderInput={params => <TextField {...params} variant="outlined" size={size} />}
    />
  );
};
