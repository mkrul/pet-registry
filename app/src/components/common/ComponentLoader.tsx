import { useEffect } from "react";
import { useAppDispatch } from "../../redux/hooks";
import { setComponentsLoading } from "../../redux/features/loading/loadingSlice";

interface ComponentLoaderProps {
  children: React.ReactNode;
}

const ComponentLoader: React.FC<ComponentLoaderProps> = ({ children }) => {
  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch(setComponentsLoading(true));

    // Use requestAnimationFrame to wait for next paint
    const timer = requestAnimationFrame(() => {
      requestAnimationFrame(() => {
        dispatch(setComponentsLoading(false));
      });
    });

    return () => {
      cancelAnimationFrame(timer);
      dispatch(setComponentsLoading(true));
    };
  }, [dispatch]);

  return <>{children}</>;
};

export default ComponentLoader;
