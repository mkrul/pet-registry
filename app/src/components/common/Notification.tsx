import React, { useEffect, useState } from "react";
import { NotificationType, NotificationProps } from "../../types/common/Notification";

const Notification: React.FC<NotificationProps> = ({ type, message, onClose }) => {
  const [isVisible, setIsVisible] = useState(true);

  useEffect(() => {
    const timer = setTimeout(() => {
      setIsVisible(false);
      setTimeout(onClose, 300);
    }, 5000);

    return () => clearTimeout(timer);
  }, [onClose]);

  if (!isVisible) return null;

  const typeStyles: Record<NotificationType, string> = {
    [NotificationType.SUCCESS]: "bg-green-100 border-green-400 text-green-700",
    [NotificationType.WARNING]: "bg-yellow-100 border-yellow-400 text-yellow-700",
    [NotificationType.ERROR]: "bg-red-100 border-red-400 text-red-700",
    [NotificationType.INFO]: "bg-blue-100 border-blue-400 text-blue-700"
  };

  return (
    <div
      className={`${typeStyles[type]} border px-4 py-3 rounded relative mb-4 flex justify-between items-center
        transition-opacity duration-300 ${isVisible ? "opacity-100" : "opacity-0"}`}
      role="alert"
    >
      <span className="block text-base font-normal sm:inline">
        {message || defaultMessages[type]}
      </span>
      <button onClick={onClose} className="ml-4" aria-label="Close">
        <span className="text-xl" aria-hidden="true">
          &times;
        </span>
      </button>
    </div>
  );
};

const defaultMessages = {
  [NotificationType.SUCCESS]: "The operation has completed successfully.",
  [NotificationType.ERROR]: "An unexpected error has occurred.",
  [NotificationType.WARNING]: "",
  [NotificationType.INFO]: ""
};

export default Notification;
