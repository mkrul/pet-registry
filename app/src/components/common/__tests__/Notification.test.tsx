import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/vitest";
import userEvent from "@testing-library/user-event";
import Notification from "../Notification";
import { NotificationType } from "../../../types/common/Notification";

describe("Notification", () => {
  const mockOnClose = vi.fn();

  beforeEach(() => {
    vi.clearAllMocks();
  });

  const renderNotification = (props = {}) => {
    return render(
      <Notification
        type={NotificationType.INFO}
        message="Test message"
        onClose={mockOnClose}
        {...props}
      />
    );
  };

  describe("Rendering", () => {
    it.each([
      {
        type: NotificationType.SUCCESS,
        className: "bg-green-100 border-green-400 text-green-700"
      },
      {
        type: NotificationType.WARNING,
        className: "bg-yellow-100 border-yellow-400 text-yellow-700"
      },
      {
        type: NotificationType.ERROR,
        className: "bg-red-100 border-red-400 text-red-700"
      },
      {
        type: NotificationType.INFO,
        className: "bg-blue-100 border-blue-400 text-blue-700"
      }
    ])("renders $type notification with correct styles", ({ type, className }) => {
      renderNotification({ type });
      const alert = screen.getByRole("alert");
      expect(alert).toBeDefined();
      expect(alert).toHaveClass(className);
    });

    it("displays custom message when provided", () => {
      const message = "Custom notification message";
      renderNotification({ message });
      expect(screen.getByText(message)).toBeDefined();
    });

    it("displays default message when no message provided", () => {
      renderNotification({
        message: "",
        type: NotificationType.SUCCESS
      });
      expect(screen.getByText("The operation has completed successfully.")).toBeDefined();
    });

    it("displays default error message for error type without message", () => {
      renderNotification({
        message: "",
        type: NotificationType.ERROR
      });
      expect(screen.getByText("An unexpected error has occurred.")).toBeDefined();
    });

    it("renders close button with correct accessibility attributes", () => {
      renderNotification();
      const closeButton = screen.getByRole("button", { name: /close/i });
      expect(closeButton).toBeDefined();
      expect(closeButton).toHaveAttribute("aria-label", "Close");
    });
  });

  describe("Interactions", () => {
    it("calls onClose when close button is clicked", async () => {
      renderNotification();
      const closeButton = screen.getByRole("button", { name: /close/i });
      await userEvent.click(closeButton);
      expect(mockOnClose).toHaveBeenCalledTimes(1);
    });
  });

  describe("Accessibility", () => {
    it("uses correct ARIA role", () => {
      renderNotification();
      expect(screen.getByRole("alert")).toBeDefined();
    });

    it("hides times symbol from screen readers", () => {
      renderNotification();
      const timesSymbol = screen.getByText("Ã—");
      expect(timesSymbol).toHaveAttribute("aria-hidden", "true");
    });
  });

  describe("Snapshots", () => {
    it.each(Object.values(NotificationType))("matches snapshot for %s notification", type => {
      const { container } = renderNotification({ type });
      expect(container).toMatchSnapshot();
    });

    it("matches snapshot with custom message", () => {
      const { container } = renderNotification({
        message: "Custom test message"
      });
      expect(container).toMatchSnapshot();
    });
  });
});
