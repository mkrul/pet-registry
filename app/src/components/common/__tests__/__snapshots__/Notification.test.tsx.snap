// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Notification > Snapshots > matches snapshot for ERROR notification 1`] = `
<div>
  <div
    class="bg-red-100 border-red-400 text-red-700 border px-4 py-3 rounded relative mb-4 flex justify-between items-center"
    role="alert"
  >
    <span
      class="block text-base font-normal sm:inline"
    >
      Test message
    </span>
    <button
      aria-label="Close"
      class="ml-4"
    >
      <span
        aria-hidden="true"
        class="text-xl"
      >
        ×
      </span>
    </button>
  </div>
</div>
`;

exports[`Notification > Snapshots > matches snapshot for INFO notification 1`] = `
<div>
  <div
    class="bg-blue-100 border-blue-400 text-blue-700 border px-4 py-3 rounded relative mb-4 flex justify-between items-center"
    role="alert"
  >
    <span
      class="block text-base font-normal sm:inline"
    >
      Test message
    </span>
    <button
      aria-label="Close"
      class="ml-4"
    >
      <span
        aria-hidden="true"
        class="text-xl"
      >
        ×
      </span>
    </button>
  </div>
</div>
`;

exports[`Notification > Snapshots > matches snapshot for SUCCESS notification 1`] = `
<div>
  <div
    class="bg-green-100 border-green-400 text-green-700 border px-4 py-3 rounded relative mb-4 flex justify-between items-center"
    role="alert"
  >
    <span
      class="block text-base font-normal sm:inline"
    >
      Test message
    </span>
    <button
      aria-label="Close"
      class="ml-4"
    >
      <span
        aria-hidden="true"
        class="text-xl"
      >
        ×
      </span>
    </button>
  </div>
</div>
`;

exports[`Notification > Snapshots > matches snapshot for WARNING notification 1`] = `
<div>
  <div
    class="bg-yellow-100 border-yellow-400 text-yellow-700 border px-4 py-3 rounded relative mb-4 flex justify-between items-center"
    role="alert"
  >
    <span
      class="block text-base font-normal sm:inline"
    >
      Test message
    </span>
    <button
      aria-label="Close"
      class="ml-4"
    >
      <span
        aria-hidden="true"
        class="text-xl"
      >
        ×
      </span>
    </button>
  </div>
</div>
`;

exports[`Notification > Snapshots > matches snapshot with custom message 1`] = `
<div>
  <div
    class="bg-blue-100 border-blue-400 text-blue-700 border px-4 py-3 rounded relative mb-4 flex justify-between items-center"
    role="alert"
  >
    <span
      class="block text-base font-normal sm:inline"
    >
      Custom test message
    </span>
    <button
      aria-label="Close"
      class="ml-4"
    >
      <span
        aria-hidden="true"
        class="text-xl"
      >
        ×
      </span>
    </button>
  </div>
</div>
`;
